### PR TITLE
Fix JSON issues

### LINKED ISSUE #3984: Bug: Json copy performance improvement
### Kùzu version

master

### What operating system are you using?

MAC M1

### What happened?

The performance of scanning ldbc1 dataset in json format is significantly slower than other systems.
After some investigation, the bottleneck lies in the 
```
common::LogicalType jsonSchema(yyjson_val* val, int64_t depth, int64_t breadth)
```
Where we try to sniff the types of each column.
Instead of reading the whole json file to sniff the datatypes, we should consider adding a sample size.

### Are there known steps to reproduce?

_No response_

### PR DESCRIPTION
**Pull request by mxwli**
Fixes bugs reported in 3947 
Fixes #3984 
Fixes performance problem related to unstructured scanning

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: extension/json/test/example.test [hunk 1]
```diff
@@ -3,42 +3,38 @@
...
 ---- ok
--STATEMENT CREATE (p:person {id: 8, description: {"family": "anatidae", "species": [ "duck", "goose", "swan", null]})
+-STATEMENT CREATE (p:person {id: 8, description: to_json({"family": "anatidae", "species": [ "duck", "goose", "swan", null]})})
 ---- ok
-#TODO(Maxwell): fix this
 -STATEMENT MATCH (p:person) return json_extract(p.description, '$.family');
 ---- 2
 
-anatidae
+"anatidae"
```
Comment (acquamarin) [inline extension/json/test/example.test:17]: Why do we have addition quotes around the string?

File: extension/json/test/example.test [hunk 2]
```diff
@@ -47,12 +43,13 @@ anatidae
...
-#-STATEMENT return cast(to_json({age: 50, names: ['Alice', 'Bob', 'Carol', 'Dan'], description: {height: 1.52}}) as struct(age int64, names string[], description struct(height int64)));
-#---- 2
-# Same query runs in duckdb:
+"alice"
+-LOG CompositeJsonStringCast
+-STATEMENT return cast(to_json({age: 50, names: ['Alice', 'Bob', 'Carol', 'Dan'], description: {height: 2}}) as struct(age int64, names string[], description struct(height int64)));
+---- 1
+{age: 50, names: ["Alice","Bob","Carol","Dan"], description: {height: 2}}
+
+# Query runs in duckdb:
```
Comment (acquamarin) [inline extension/json/test/example.test:52]: Removes the comment

File: extension/json/test/example.test [hunk 3]
```diff
@@ -3,42 +3,38 @@
...
 -CASE ExtractNotExistField
--SKIP
 -STATEMENT LOAD EXTENSION "${KUZU_ROOT_DIRECTORY}/extension/json/build/libjson.kuzu_extension";
 ---- ok
 -STATEMENT create node table person (id INT64, description json, primary key(id));
 ---- ok
 -STATEMENT CREATE (p:person {id: 5, description: 6})
 ---- ok
--STATEMENT CREATE (p:person {id: 8, description: {"family": "anatidae", "species": [ "duck", "goose", "swan", null]})
+-STATEMENT CREATE (p:person {id: 8, description: to_json({"family": "anatidae", "species": [ "duck", "goose", "swan", null]})})
```
[the remark this answers is not in the document]
Comment (mxwli) [reply extension/json/test/example.test:12]: We can, but not here. If I didn't write to_json, the description column would be implicitly casted to STRING. I'm fairly certain we only have the necessary information to determine the target type is JSON when we have something like CAST(1 AS JSON)

File: extension/json/test/example.test (hunk 1 above)
Comment (mxwli) [reply extension/json/test/example.test:17]: JSON strings must have double quotes around them.

File: extension/json/src/json_utils.cpp [hunk 4]
```diff
@@ -606,45 +606,48 @@ JsonWrapper stringToJsonNoError(const std::string& str) {
...
+static JsonWrapper fileToJsonUnstructuredFormatted(std::shared_ptr<char[]> buffer,
+    uint64_t fileSize) {
     JsonMutWrapper result;
     auto root = yyjson_mut_arr(result.ptr);
     yyjson_mut_doc_set_root(result.ptr, root);
     auto filePos = 0u;
     while (true) {
-        auto curDoc = yyjson_read(buffer + filePos, fileSize - filePos, YYJSON_READ_STOP_WHEN_DONE);
+        auto curDoc = yyjson_read_opts(buffer.get() + filePos, fileSize - filePos,
+            YYJSON_READ_STOP_WHEN_DONE | YYJSON_READ_INSITU, nullptr, nullptr);
```
Comment (acquamarin) [inline extension/json/src/json_utils.cpp:625]: Add comment for nullptr

File: extension/json/src/json_utils.cpp [hunk 5]
```diff
@@ -606,45 +606,48 @@ JsonWrapper stringToJsonNoError(const std::string& str) {
...
+    return JsonWrapper(yyjson_mut_doc_imut_copy(result.ptr, nullptr), buffer);
 }
 
 JsonWrapper fileToJson(main::ClientContext* context, const std::string& path,
     JsonScanFormat format) {
 
     auto file = context->getVFSUnsafe()->openFile(path, O_RDONLY, context);
     auto fileSize = file->getFileSize();
-    std::unique_ptr<char[]> buffer(new char[fileSize + 4]());
+    std::shared_ptr<char[]> buffer(new char[fileSize + 9]());
```
Comment (acquamarin) [inline extension/json/src/json_utils.cpp:641]: use `auto buffer = std::make_shared<char[]>(fileSize + 9)`

File: extension/json/src/json_utils.cpp [hunk 6]
```diff
@@ -606,45 +606,48 @@ JsonWrapper stringToJsonNoError(const std::string& str) {
...
+    memset(buffer.get() + fileSize, 0, 9);
```
Comment (acquamarin) [inline extension/json/src/json_utils.cpp:642]: Add comment for constants 0 and 9

File: extension/json/src/json_utils.cpp [hunk 7]
```diff
@@ -655,7 +658,7 @@ JsonWrapper mergeJson(const JsonWrapper& A, const JsonWrapper& B) {
     JsonMutWrapper ret;
     auto root = yyjson_merge_patch(ret.ptr, yyjson_doc_get_root(A.ptr), yyjson_doc_get_root(B.ptr));
     yyjson_mut_doc_set_root(ret.ptr, root);
-    return JsonWrapper(yyjson_mut_doc_imut_copy(ret.ptr, nullptr));
+    return JsonWrapper(yyjson_mut_doc_imut_copy(ret.ptr, nullptr), nullptr);
```
Comment (acquamarin) [inline extension/json/src/json_utils.cpp:661]: Add comment for nullptr

File: extension/json/test/example.test (hunk 3 above)
[the remark this answers is not in the document]
Comment (acquamarin) [reply extension/json/test/example.test:12]: Why can duckdb allow user to write json like:
```
insert into person values(8, '{"family": "anatidae", "species": [ "duck", "goose", "swan", null]}');
```?

File: src/function/vector_cast_functions.cpp [hunk 8]
```diff
@@ -1262,11 +1262,25 @@ static std::unique_ptr<FunctionBindData> castBindFunc(ScalarBindFuncInput input)
     }
     auto literalExpr = input.arguments[1]->constPtrCast<LiteralExpression>();
     auto targetTypeStr = literalExpr->getValue().getValue<std::string>();
+    auto func = input.definition->ptrCast<ScalarFunction>();
     // TODO(Ziyi): we should pass the clientContext pointer here so the bind function can access
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:1266]: remove this comment

File: src/function/vector_cast_functions.cpp [hunk 9]
```diff
@@ -1262,11 +1262,25 @@ static std::unique_ptr<FunctionBindData> castBindFunc(ScalarBindFuncInput input)
     }
     auto literalExpr = input.arguments[1]->constPtrCast<LiteralExpression>();
     auto targetTypeStr = literalExpr->getValue().getValue<std::string>();
+    auto func = input.definition->ptrCast<ScalarFunction>();
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:1265]: Move this variable into the if statement that you just wrote below

File: src/function/vector_cast_functions.cpp [hunk 10]
```diff
@@ -1262,11 +1262,25 @@ static std::unique_ptr<FunctionBindData> castBindFunc(ScalarBindFuncInput input)
...
     // TODO(Ziyi): we should pass the clientContext pointer here so the bind function can access
     // the user defined types.
     LogicalType targetType;
     if (!LogicalType::tryConvertFromString(targetTypeStr, targetType)) {
         targetType = input.context->getCatalog()->getType(input.context->getTx(), targetTypeStr);
+        std::vector<LogicalType> typeVec;
```
Comment (acquamarin) [inline src/function/vector_cast_functions.cpp:1271]: Refactor the whole if statement as a function called `castUDT`

File: extension/json/src/include/json_utils.h [hunk 11]
```diff
@@ -16,15 +16,18 @@ class JsonWrapper {
...
     ~JsonWrapper();
     JsonWrapper(JsonWrapper& other) = delete;
     JsonWrapper(JsonWrapper&& other) {
         ptr = other.ptr;
         other.ptr = nullptr;
+        buffer = std::move(other.buffer);
     }
 
     yyjson_doc* ptr;
+    std::shared_ptr<char[]> buffer;
```
Comment (acquamarin) [inline extension/json/src/include/json_utils.h:30]: our convension is to use std::shared_ptr<uint8[]>

File: extension/json/src/json_scan.cpp [hunk 12]
```diff
@@ -74,8 +74,19 @@ struct JsonBindData : public ScanBindData {
             scanFromList, scanFromStruct, config.copy(), context);
     }
 
-    uint32_t getIdxFromName(const std::string& s) const {
-        return nameToIdxMap.at(StringUtils::getUpper(s));
+    int32_t getIdxFromName(std::string s) const { // possible bottleneck
+        auto cpy = StringUtils::getUpper(s);
+        if (!nameToIdxMap.contains(cpy)) {
+            // try removing any [a-zA-Z]+\. prefix
```
Comment (acquamarin) [inline extension/json/src/json_scan.cpp:80]: In which case, the name would contain a prefix? Can you add a comment there?

File: extension/json/src/json_scan.cpp [hunk 13]
```diff
@@ -74,8 +74,19 @@ struct JsonBindData : public ScanBindData {
...
+            auto periodPos = cpy.find('.');
+            if (periodPos != std::string::npos) {
+                cpy = cpy.substr(periodPos + 1);
+            }
+        }
+        if (!nameToIdxMap.contains(cpy)) {
+            return -1;
```
Comment (acquamarin) [inline extension/json/src/json_scan.cpp:87]: Should we throw an exception if we don't find the field name

**Review by acquamarin — APPROVED** [review]

**Pull request merged**